src/khoj/database/adapters/__init__.py

```diff
@@ -405,6 +406,32 @@ async def aget_user_search_model(user: KhojUser):
     return config.setting
 
 
+class ProcessLockAdapters:
+    @staticmethod
+    def get_process_lock(process_name: str):
+        return ProcessLock.objects.filter(name=process_name).first()
+
+    @staticmethod
+    def set_process_lock(process_name: str, max_duration_in_seconds: int = 600):
+        return ProcessLock.objects.create(name=process_name, max_duration_in_seconds=max_duration_in_seconds)
+
+    @staticmethod
+    def is_process_locked(process_name: str):
+        process_lock = ProcessLock.objects.filter(name=process_name).first()
+        if not process_lock:
+            return False
+        if process_lock.started_at + timedelta(seconds=process_lock.max_duration_in_seconds) < datetime.now(
+            tz=timezone.utc
+        ):
+            process_lock.delete()
+            return False
+        return True
+
+    @staticmethod
+    def remove_process_lock(process_name: str):
+        return ProcessLock.objects.filter(name=process_name).delete()
```

Comment:
Not sure, but wondering if instead of creating and deleting the process locks we should update the state of the process locks instead?

This will allow other background processes to gather from the process lock DB table:
1. When a specific operation last ran (e.g via a new `completed_at` field in table)
2. If the previous operation run failed or succeeded (e.g via new `operation_state` field in table with new, started, failed, succeeded choices)

Reply:
> When a specific operation last ran (e.g via a new completed_at field in table)

Hmm. I think this would be moving away from the purpose of acting as a db lock, and more moving into a role as a job manager. I think we should take care to avoid conflating those concepts. If we have a separate place where we manage `job`s generally, we should definitely add metadata like that in there. 

> If the previous operation run failed or succeeded (e.g via new operation_state field in table with new, started, failed, succeeded choices)

If there's a DB lock with a certain name present in the table, we implicitly have this information. If it's started, the process lock is there. If it's failed (without the process exiting), the lock would be removed. If it's succeeded, the lock would be removed. 

The tricky thing is when it fails silently, leaving us in a deadlock situation until the `max_duration_in_seconds` kicks in.